src/cryptonote_basic/tx_extra.h

```diff
@@ -235,20 +254,74 @@ namespace cryptonote
     END_SERIALIZE()
   };
 
-  struct tx_extra_service_node_deregister
+  struct tx_extra_service_node_state_change
   {
     struct vote
     {
       vote() = default;
-      vote(crypto::signature const &signature, uint32_t validator_index): signature(signature), validator_index(validator_index) { }
+      vote(crypto::signature signature, uint32_t validator_index) : signature{std::move(signature)}, validator_index(validator_index) { }
       crypto::signature signature;
       uint32_t          validator_index;
+
+      BEGIN_SERIALIZE()
+        VARINT_FIELD(validator_index);
+        FIELD(signature);
+      END_SERIALIZE()
     };
 
+    service_nodes::new_state state;
+    uint64_t                 block_height;
+    uint32_t                 service_node_index;
+    std::vector<vote>        votes;
+
+    tx_extra_service_node_state_change() = default;
+
+    template <typename... VotesArgs>
+    tx_extra_service_node_state_change(service_nodes::new_state state, uint64_t block_height, uint32_t service_node_index, VotesArgs &&...votes)
+        : state{state}, block_height{block_height}, service_node_index{service_node_index}, votes{std::forward<VotesArgs>(votes)...} {}
+
+    // Compares equal if this represents a state change of the same SN (does *not* require equality of stored votes)
+    bool operator==(const tx_extra_service_node_state_change &sc) const {
+      return state == sc.state && block_height == sc.block_height && service_node_index == sc.service_node_index;
```

Comment:
rewrite in terms of `std::tie` (and make a free function):

```c++
bool operator==(const tx_extra_service_node_state_change &lhs, 
                const tx_extra_service_node_state_change &rhs)
{
    return std::tie(lhs.state, lhs.block_height, lhs.service_node_index) == 
           std::tie(rhs.state, rhs.block_height, rhs.service_node_index);
}
```

Reply:
What's the practical advantage of doing these two things?

Reply:
I guess it looks a little cleaner, but I'm kind of ambivalent about it.  I'm wondering more about the advantage of having it a free function rather than a member function.